Articles/Blog/2024-05-communicating-design.md

```diff
@@ -0,0 +1,133 @@
+- In a few sentences, what specifically are you communicating?
+- At what level does the content of the message exist in the software: conceptual, specification, or implementation?
+
+The three diagrams in Figure 3 are taken from one of my software projects, FLORIS, a wind farm wake modeling framework that provides specific interfaces where developers can plug in new wake models.
+Consider the three perspectives:
+
+- *Conceptual* describes the relationships between the main components of the software and notes in red where to connect a new wake model.
+- *Specification* describes the connections between a portion of the wake model and the software objects that interface with it.
+- *Implementation* describes the specific attributes on a particular class and its inherited properties.
+
+While each is valuable, maintaining separation allows for focusing a diagram on specific themes for a given audience.
+
+<img src='../../images/Blog_2024_perspectives.png' class='page lightbox'/>[Figure 3: A comparison of conceptual, specific, and implementation perspectives in diagrams from the FLORIS framework.]
+
+### Documentation-driven development
+
+As in test-driven development, [documentation-driven development](https://www.writethedocs.org/videos/portland/2019/lessons-learned-in-a-year-of-docs-driven-development-jessica-parsons/)
+is the practice of stating what the code should do prior to writing it,
+and diagramming at this stage can make the design process more effective.
+The result of documentation-driven development may be a design document that can take the form
+of an internal document, a page in online documentation, or a GitHub Discussion or Issue.
+Suggested content to include in a design document are:
+
+- Scope and suggested design of the work
+- Relationship of new code to existing elements of the software including current implementations and overarching themes
+- New themes and design decisions included and excluded
+
+While a strict policy of "docs or it didn't happen" can increase the quality and quantity of documentation, it comes with the added burden for developers, reviewers, and maintainers.
+A more approachable process is to require an extended pull request description that includes a narrative of the changes and an overview of design decisions.
+In any case, including automated diagramming tools in the software infrastructure can reduce the burden
+and empower developers to adopt this practice.
+
+### Summary
+
+Through the BSSw Fellowship, I've had the opportunity to gather and formalize ideas on documentation and communication on software design.
+I presented at the NLIT S3C conference in April 2024 ([slides](https://rafmudaf.github.io/communicating-design/_downloads/67486fd27e6ced8dd8672408a18de874/nlit_s3c.pdf))
+and will present an [HPC Best Practices webinar](https://ideas-productivity.org/resources/series/hpc-best-practices-webinars/) in June 2024.
```

Comment:
@bernhold should this link be replaced with the one for the webinar?
https://ideas-productivity.org/events/hpcbp-084-communicatingdesign

Reply:
Yeah, please do that.

Reply:
Done